Generators (#270)
javascript-tutorial/fa.javascript.info: 1-js/12-generators-iterators/1-generators/article.md

# Generators

> توابع معمول فقط یک‌بار می‌توانند یک مقدار(مثلا یک object یا undefined) را برگردانند.

Reviewer: ```suggestion
توابع معمولی فقط یک‌بار می‌توانند یک مقدار(مثلا یک object یا undefined) را برگردانند.
```

> اما generatorها می‌توانند چندین بار، بر اساس تقاضا، مقادیر متفاوت را برگردانند(اصطلاحا yield کنند.). generatorها با [iterableها](info:iterable) به خوبی کار می‌کنند و به کمک آن‌ها می‌توان جریان‌های داده ساخت.

Reviewer: ```suggestion
اما generatorها می‌توانند چندین بار، بر اساس تقاضا، مقادیر متفاوت را برگردانند(اصطلاحا yield کنند.). generatorها با [حلقه‌پذیرها](info:iterable) (iterable) به خوبی کار می‌کنند و به کمک آن‌ها می‌توان جریان‌های داده ساخت.
```

}
```

> توابع generator با توابع معمول، رفتار متفاوتی دارند. زمانی که این توابع صدا می‌شوند، بدنه آن‌ها اجرا نمی‌شود؛ در عوض، یک آبجکت خاص به نام "generator object" برمی‌گردانند که به وسیله آن اجرای تابع را می‌توان کنترل کرد.

Reviewer: ```suggestion
توابع generator با توابع معمولی، رفتار متفاوتی دارند. زمانی که این توابع صدا می‌شوند، بدنه آن‌ها اجرا نمی‌شود؛ در عوض، یک شیء خاص به نام "generator object" برمی‌گردانند که به وسیله آن اجرای تابع را می‌توان کنترل کرد.
```

  return 3;
}

> //برمی‌گرداند generator یک آبجکت generator تابع

Reviewer: ```suggestion
//برمی‌گرداند generator یک شیء generator تابع
```

let generator = generateSequence();
\*!*
alert(generator); // [object Generator]

> \*/!*

Reviewer: لطفا دلیل اضافه کردن این دوتا رو بگین.

```

اجرای بدنه تابع هنوز شروع نشده است:

![](generateSequence-1.svg)

> متد اصلی یک آبجکت generator متد `()next` است. هنگامی که صدا می‌شود، بدنه تابع تا اولین `yield value` اجرا می‌شود(`value` می‌تواند حذف شود که در این صورت `undefined` است.)؛ سپس اجرای تابع متوقف می‌شود و مقدار yield شده برگرداننده می‌شود.

Reviewer: ```suggestion
متد اصلی یک شیء generator متد `()next` است. هنگامی که صدا می‌شود، بدنه تابع تا اولین `yield value` اجرا می‌شود(`value` می‌تواند حذف شود که در این صورت `undefined` است.)؛ سپس اجرای تابع متوقف می‌شود و مقدار yield شده برگرداننده می‌شود.
```
لطفا "object" رو به «شیء» ترجمه کنین :)

مقدار برگردانده شده توسط متود `next` همواره یک آبجکت با 2 پراپرتی است:
- `value`: مقدار برگرداننده شده توسط `yield`.
- `done`: یک Boolean است که در صورت اتمام بدنه تابع مقدار true و در غیر این صورت مقدار false دارد.

> برای مثال، در کد زیر، یک آبجکت generator ایجاد شده و اولین مقدار `yield` شده توسط آن گرفته شده است:

Reviewer: ```suggestion
برای مثال، در کد زیر، یک شیء generator ایجاد شده و اولین مقدار `yield` شده توسط آن گرفته شده است:
```

![](generateSequence-4.svg)

اکنون از روی `done:true` متوجه می‌شویم کار generator تمام شده و `value:3` آخرین مقدار برگردانده شده توسط generator است.

> دیگر صدا کردن `()generator.next` منطقی نیست. اگر این کار را انجام دهیم، آبجکت یکسانی با `done:true` برگردانده می‌شود.

Reviewer: ```suggestion
دیگر صدا کردن `()generator.next` منطقی نیست. اگر این کار را انجام دهیم، شیء یکسانی با `done:true` برگردانده می‌شود.
```

```smart header="`function* f(…)`یا`function *f(…)`؟"
هر دو سینتکس صحیح هستند.

ولی معمولا اولی ترجیح داده می‌شود؛ چون `*` نوع تابع و نه نام تابع را مشخص می‌کند.
````

> ## generatorها iterable هستند.

Reviewer: ```suggestion
## generatorها حلقه‌پذیر هستند.
```

> همان‌طور که احتمالا با توجه به `()next` متوجه شده‌اید، generatorها [iterable](info:iterable) هستند.

Reviewer: ```suggestion
همان‌طور که احتمالا با توجه به `()next` متوجه شده‌اید، generatorها [حلقه‌پذیر](info:iterable) هستند.
```

let generator = generateSequence();

for(let value of generator) {
  alert(value); // ابتدا 1 و سپس 2
}
````

این شیوه از صدا کردن `next` تمیزتر است؛ این‌گونه فکر نمی‌کنید؟

...اما دقت کنید: مثال بالا ابتدا `1` و سپس `2` را نشان داد؛ خبری از `3` نیست!

> علت این اتفاق این است که `for..of` آخرین مقدار را هنگامی که `done:true` است در نظر نمی‌گیرد. هنگام برگرداندن آخرین مقدار با `return`، بر خلاف `done:true`، `yield` است. در نتیجه برای نشان دادن تمام مقادیر باید آن‌ها را با `yield` برگردانیم:

Reviewer: ```suggestion
علت این اتفاق این است که `for..of` آخرین مقدار را هنگامی که `done:true` است در نظر نمی‌گیرد. آخرین مقدار برگردانده شده با `return`، بر خلاف `yield`، حاوی `done:true` است. در نتیجه برای نشان دادن تمام مقادیر در حلقه `for..of` باید آن‌ها را با `yield` برگردانیم:
```

```js run
function* generateSequence() {
  yield 1;
  yield 2;
\*!*
  yield 3;
\*/!*
}

let generator = generateSequence();

for(let value of generator) {
  alert(value); // ابتدا 1 سپس 2 و بعد از آن 3
}
```

> از آنجایی که generatorها iterable هستند، از تمام functionality آن‌ها نیز برای generatorها می‌توان استفاده کرد؛ مثل spread syntax `...`:

Reviewer: ```suggestion
از آنجایی که generatorها حلقه‌پذیر هستند، از تمام functionality آن‌ها نیز برای generatorها می‌توان استفاده کرد؛ مثل spread syntax `...`:
```

alert(sequence); // 0, 1, 2, 3
```

> در کد بالا، `()generateSequence...`، باعث می‌شود آبجکت generator که iterable هم هست به آرایه‌ای از اعداد تبدیل شود.(درباره spread syntax در چپتر [](info:rest-parameters-spread#spread-syntax) بیشتر بخوانید.)

Reviewer: ```suggestion
در کد بالا، `()generateSequence...`، باعث می‌شود شیء generator که حلقه‌پذیر هم هست به آرایه‌ای از اعداد تبدیل شود.(درباره spread syntax در فصل [](info:rest-parameters-spread#spread-syntax) بیشتر بخوانید.)
```

> ## استفاده از generatorها برای iterableها

Reviewer: ```suggestion
## استفاده از generatorها برای حلقه‌پذیرها
```

alert(str); // 0..9A..Za..z
```

ترکیب generatorها یک راه معقول برای استفاده از جریان یک generator درون دیگری است و از حافظه بیشتر برای ذخیره مقادیر استفاده نمی‌کند.

## yield یک خیابان دو طرفه است.

> تا الان، generatorها بسیار شبیه به آبجکت‌های iterable با یک سینتکس خاص برای تولید مقادیر بودند؛ درواقع اما generatorها بسیار قدرتمندتر و انعطاف پذیرتر هستند.

Reviewer: ```suggestion
تا الان، generatorها بسیار شبیه به شیءهای حلقه‌پذیر با یک سینتکس خاص برای تولید مقادیر بودند؛ درواقع اما generatorها بسیار قدرتمندتر و انعطاف پذیرتر هستند.
```

چون `yield` یک خیابان دو طرفه است: نه تنها مقدار را به بیرون برمی‌گرداند بلکه می‌تواند مقادیر را به داخل generator بیاورد.

> برای این کار باید `()generator.next` را با یک argument صدا کنیم. این argument تبدیل به مقدار برگردانده شده توسط خود yield درون generator می‌شود.

Reviewer: ```suggestion
برای این کار باید `()generator.next` را با یک آرگومان صدا کنیم. این argument تبدیل به مقدار برگردانده شده توسط خود yield درون generator می‌شود.
```

const g = gen();

g.next(); // { value: 1, done: false }
g.return("foo"); // { value: "foo", done: true }
g.next(); // { value: undefined, done: true }
```

اگر در یک generatorخاتمه یافته دوباره از `()generator.return`استفاده کنیم، همان مقدار را دوباره برمی‌گرداند.([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return)).

معمولا از آن استفاده نمی‌شود؛ چون اکثر زمان‌ها می‌خواهیم تمام مقادیر را بگیریم، اما موقعی که بخواهیم در شرایط خاص generator را متوقف کنیم کاربرد دارد.

## خلاصه

- generatorها توسط تابع generator تولید می‌شوند. `{...} (...) function* f`
- عملگر yield (فقط) در داخل generatorها وجود دارد.
- کد بیرونی و generator ممکن است توسط `next/yield` با هم نتایج را رد و بدل کنند.

> در جاوااسکریپت مدرن، generatorها کم استفاده می‌شوند. اما گاهی اوقات می‌توانند مفید باشند؛ رد و بدل کردن داده با کد صدا زننده، یک قابلیت منحصر بفرد است. هم‌چنین برای ساخت iterableها هم بکار می‌روند.

Reviewer: ```suggestion
در جاوااسکریپت مدرن، generatorها کم استفاده می‌شوند. اما گاهی اوقات می‌توانند مفید باشند؛ رد و بدل کردن داده با کد صدا زننده، یک قابلیت منحصر بفرد است. هم‌چنین برای ساخت حلقه‌پذیرها هم بکار می‌روند.
```
